### DISCUSSION TITLE
第五回 MTG (9/27)

### DISCUSSION THREAD

**Discussion by MachineHunter** [MachineHunter/SafeExecute#49] [General]
## 話し合いたい内容
* (ここまでで機能実装はほぼ終了しているという前提)
* 主に評価やバグに関して話し合う（一応確認ですがrules.csvのWinAPI全部1でちゃんと評価しました？）
* ↑の上で、改めてこのツールの売り、どう魅せるかの部分を考えていく
* また、最低限追加で実装しないといけない点や修正箇所を相談する
* リリースについて話し合う（githubのreleasesでzip形式で一つのフォルダに詰め込んで、相対パスでいい気がしてきた）
* そういえばアプリの名前どうしましょうか

## 成果発表に関して
提出するのは、以下
* ツール、ソースコード、**マニュアル等のドキュメント**
  * 「多くの人が利用や改良可能にしてください」とも書いてある
  * 継続性も評価の枠の一つ
* **プレゼン資料**（PDFファイル）
* **プレゼン動画**（mp4ファイル）

プレゼン資料には、特に指定は無いが、以下が軸になると思う
* 頑張った点（実装の難しかった点など）
* アピールすべき細かい工夫/配慮
* ツールの使いどころ（このツール実際使えるのか、意味があるのか）

## セルフ評価
* 新規性
  * ビミョイ
  * 個人で実行をカスタマイズできるという点がコアな部分で新規性があるが、技術的にはEDR製品と似てる
  * => もうちょい何か個性が欲しい
* 実用性
  * 評価中
  * 誤検知はほぼ無さげでグッド（[wiki 通常アプリの評価](https://github.com/MachineHunter/SafeExecute/wiki/通常アプリの評価)）
  * 現在の脅威に対して有効かどうかはまだチェックできてない
    * VM持ってる人にTrojan系のマルウェア中心に試してほしい
* 継続性
  * 未着手
  * 今後実装すべきものもissueに書いていきましょう～
  * ドキュメントとかプレゼン資料そろそろ作り始めましょう！
* チームワーク
  * かなり良さげ
  * issueのAssignee、PRを見ればよくわかる
  * 後はissueをプレゼンでアピールすればいい

## 追加機能実装
個人的な考えとしては、
* インターネット接続のAPIこの二種類(WinSock, WinInet) 以外にもありそう
* マルウェアの永続化手法がRUNレジストリとWinサービス以外にもある気がする
* もう一つくらい大きいジャンルが欲しい
* ファイルダウンロードの検知が欲しい

**Comment by MachineHunter** [comment]
# 予定

## 機能実装
* 9/30まで（できれば早めに）
* 1人 1マルウェア を解析レポートとか見て調査し、それに関連する検知を1つ実装する
  * マルウェアについて調べたら、それをwikiにまとめてほしい
* 「脅威をちゃんと検知できるか」を評価するために、実際にマルウェアとかでSafeExecutorを試してみる
  * VMがある人に限る（マルウェア動かしてみたいという人がいれば是非）
  * 10/3までを目安に @KOHsEMP 君にやってもらえれば！


## 評価
* 10/6には終わらせたい（10/7 23:59 〆切）
* ドキュメント化した方がいい物とかあれば、ここのDiscussionsのコメントとかに書いていただいて、他の人と被らないように上手く調整しながらやっていってほしい
  * issueのClosedを見てほしい
  * 他の人がPRとかを作成する場合に、知っておかないといけない点をどこかで公開

※ @kbys0519 君: 9/30まで予定が密なため、10/1, 2 で機能実装に入ってもらう形になりそうだが、リリースとかの関係も含めて実際どうするかは後で決める

**Comment by MachineHunter** [comment]
## アプリの名前
**「安全実行侍\~俺を信じろ\~」** に決定！

**Comment by MachineHunter** [comment]
## この ツール/プロジェクト の 売り/見せ方

### ITとかそこまで詳しくないユーザーへの売り
* issueでチームワークを見てほしい
* C++もWinAPIもマルウェアも初めてのチームで作った
* **実行ファイルの挙動をユーザーが制御できる**というのがメインで新規性の部分
* ユーザーが実行を制御できる
  * バックグラウンドで常時起動してるわけではなく、オンデマンドで実行できる
* 正常なバイナリだった場合に、正常な実行に全く支障が無いです
* 軽い
  * アンチウイルスだとアップデートで処理が重くなったりするけど、そういうのに比べて軽量（容量）
* アンチウイルスだと動かしたいプログラムを勝手に消されたりするが、そんなことはしません！
* 実行ファイルがどこと通信してるのかがわかる

### マルウェア解析者やITに詳しい方にとってのメリット
* マルウェア解析者にとってのメリット
  * 手軽に、ある特定のWinAPIがどういう引数で呼ばれてるのかを確認できる
  * その呼ばれてる時点でメッセージボックスが表示されるので、そのタイミングでデバッガをアタッチできる
  * 検知したWinAPIのログを学習用データに使える

**Comment by MachineHunter** [comment]
## ドキュメント案
* README.md
* README_en.md
* wiki
* どうやってContributeするか

**Comment by MachineHunter** [comment]
## マルウェアのトレンド
何のマルウェア見ていいかわからないという方は、
最近増えてるマルウェアを見てみるといいかもです。
https://any.run/malware-trends/

ランキング的にはこんな感じです
1. Redline
2. Agenttesla
3. Snake
4. Formbook
5. Emotet
6. Qbot
7. Asyncrat
8. Raccoon
9. Njrat
10. Nanocore

あ、後マルウェアといっても種類がありますが、できれば Trojan系 がいいです。
Trojanは正規のアプリに見せかけて実はマルウェア ってやつで、このツールはそういうのを対象にしてるので。

最初はマクロで感染するマルウェアがあるので、そういうのは対象外にするように注意です

**Reply by kbys0519** [reply]
自分はランサムウェアRagnar Lockerについて検知を実装します！

**Reply by KOHsEMP** [reply]
ランサムウェアMazeについてまずは検知とWikiを書きます！

**Reply by KOHsEMP** [reply]
SnakeについてWikiを書きます！

**Comment by shhu6844** [comment]
# SafeExecute (English Version)

## Summary
### SafeExecute is an executable checker that will detect any Windows API calls during runtime.
In detail, we implement our API hooking code into a dll file called "SafeExecute.dll". When users provide their executables, SafeExecute will run it with a dll file which changes the behavior of the target executable into multiple confirmation messages when any chosen Windows API is called.

## Usage
### SafeExecute provides safe execution to some unknown executable files and notify users with all of the Windows API calls during their runtime. 
User is able to either continue it or interrupt it when any hook occurs. It's also possible to learn what API has been used on any applications you want to test with SafeExecute!

## Features
* Runtime Analysis
* Portable
* Self-Determination
* Understand Program Content
* User-Friendly Suggestion System

Compared to Anti-Malware or Anti-Virus system which often scan and detect executables in your computer, SafeExecute is a portable tool that does dynamic detection without using any dataset and Internet to support its functioning. For instance, Anti-Malware software may able to erase files when detect something malicious, however, the software does not tell the reason being suspicious, so if user wants to learn more about some malicious files, SafeExecute is the best choice that protect user by assisting user during the execution and provide an option to interrupt the target executables when found suspicious Windows API being called, also they can decide whether to continue or not by themselves, without any third-party's suggestions or judgements.
It's also a good tool for experienced researchers and programmers to investigate Windows executables!

## Installation(For people who don't have Visual Studio developing environment)
* If you already have Visual Studio in your computer, just open "SafeExecutorGUI.exe".
* If you don't have Visual Studio in your computer, please follow the steps below to open it with MSIX package. We need users to add Reliable Creator to your computer manually.
  1. Right click ".crt" file and choose **Install Certificate**.
  2. You can choose to install it on local computer.
  3. After installing, you can check if this certificate is in the list by opening "certmgr. msc" from Windows Search Engine beside your Windows icon. (Default is at left bottom of your screen)
  4. Double click ".appxbunldle" file and start install SafeExecute.

## How to...
1. Open "SafeExecutorGUI.exe"
2. Select your target executable file.
3. If any argument is needed, you can input them in the second line.
4. Choose Mode. There are Normal mode with stdout, Error mode with stderr, and Interactive mode with messagebox showing, Interactive mode is suggested.
5. Click on the Windows API type you interested in.
6. Click "Run" and start!

For example:
```sh
echo.exe "arg1"
```
After selecting echo.exe, you need to put the argument in the arguments section.

## Important Files
* **SafeExecute.dll**: Contains our hook methods, Windows API will be detected by our code and will not execute without the user's confirmation.
* **SafeExecutor.exe**: A program combining hook methods and target executables with arguments, we can consider it as an input machine. Users can just use "SafeExecutorGUI.exe" when using!
* **SafeExecutorGUI.exe**: GUI interface for SafeExecute.

## Command Line Usage
```sh
./SafeExecutor.exe SafeExecute.dll <path-to-executables>
```

## Showcase

```sh
Should have "GIF files" here!
```
Here is an example of executing a program that wants to open a url of Github. And we tried to allow all the APIs by clicking "yes".
![](./img/InternetExample01.gif)
And here is the part when we reject its execution.
![](./img/InternetExample02.gif)
Here is an example of executing a fake ransomware we made that tries to encrypt all the ".txt" files in the path of input argument. And we all the APIs by clicking "yes".
![](./img/FakeRansom01.gif)
And here is the part when we reject its execution.
![](./img/FakeRansom02.gif)

## Behind the Scenes
SafeExecute uses an API hook to monitor target executable. Users are able to decide whether to continue or not when SafeExecute hooks any Windows API user select.
About how we produced this technique, here is an example code we programmed similar to our released code. Let's see through this example executable "hello.exe":

```c++
#include <windows.h>
#include <cstdio>

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow) {
    MessageBoxA(NULL, "Hello World!", "Info", MB_YESNO);
    return 0;
}
```

This program is really simple. It'll show a message box by calling Windows API "MessageBoxA", and this API is from library "user32.dll", so the process during runtime will look like the image below.

```sh
Should have "./img/callMessageBoxA.png" here!
```
![](./img/callMessageBoxA.png)

About what we try to do in SafeExecute is that instead of getting the real "MessageBoxA", we covered the address of "MessageBoxA_Hook" ,which is our contribution in SafeExecute.dll, over the address of "MessageBoxA". The process during runtime will become like the image below.

```sh
Should have "./img/callHookedMessageBoxA.png" here!
```
![](./img/callHookedMessageBoxA.png)

You may somehow wonder what is inside "MessageBoxA_Hook", before calling the real "MessageBoxA" we added a message box that contains "yes" and "no" with some knowledge about the function calls being hooked, in that case users can figure out whether to continue or not by themselves, without any third-party's suggestions or judgements. If the user chooses not to continue, SafeExecute will easily call "ExitProcess" at that moment and end the process.

In many cases, SafeExecute will hook out many Windows APIs because of the wide variety of functionalities in a program or executable. You can still continue running by clicking "yes" every time a hook pops up, to make it personalize we also added a checklist which users can choose what kind of Windows API should be hooked, and the others will not.

I hope you enjoy using SafeExecute. Good Luck!
